keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,449 @@
+  // Workload.
+  // preemption is immutable once set.
+  Preemption *WorkloadPreemption
+}
+
+type WorklodPreemption struct {
+  // disruptionCostMilestones defines a cost curve for preempting this workload
+  // based on the time the workload has been running.
+  // The curve is defined by points of (time, cost). The cost is calculated from
+  // a linear interpolation of the cost between the two time points that contain
+  // the current workload runtime.
+  // The first point is implicitly (0, 0)
+  // The cost from the last milestone is constant.
+  //
+  // For example: given the points (60, 0) and (120, 100), the cost of
+  // preempting the Workload after it has been running for:
+  // - 30s is 0,
+  // - 90s is 50,
+  // - 300s is 100.
+  //
+  // disruptionCostMilestones[i].time > disruptioncostMilestones[j].time for i > j
+  DisruptionCostMilestones []DisruptionCostMilestone.
+}
+
+type DisruptionCostMilestone struct {
+  // seconds is the time since the Workload started (timestamp of the Started
+  // condition on the Workload).
+  Seconds int64
+
+  // cost is an integer representation of the cost of disrupting the Workload
+  // at a given time. The value has meaning when compared to the cost of
+  // disrupting other Workloads.
+  //
+  // 0 <= cost <= 100
+  Cost int32
+}
+```
+
+#### LocalQueue changes
+
+The new API fields in LocalQueue describe the defaults that apply to Workloads
+in this LocalQueue.
+
+```golang
+type LocalQueueSpec struct {
+  ...
+  // preemptionDefaults are the defaults applied to Workloads assigned to this
+  // LocalQueue. The defauts are applied to Workloads at the time of the
+  // Workload admission.
+  PreemptionDefaults *WorkloadPreemption
+}
+```
+
+#### ClusterQueue changes
+
+The new API fields in ClusterQueue describe how to influence the selection
+of Workloads to preempt.
+
+```golang
+const ClusterQueuePendingWorkloadsUnderMinQuota = "PendingWorkloadsUnderMinQuota"
+
+type ClusterQueueSpec struct {
+  ...
+  // preemption describes policies to preempt Workloads from ClusterQueues
+  // borrowing resources when this ClusterQueue has pending Workloads and the
+  // min quota is unmet.
+  preemption ClusterQueuePreemption
+}
+
+type ClusterQueueStatus struct {
+  ...
+  conditions []metav1.Condition
+}
+
+type ClusterQueuePreemption struct {
+  // waitBeforePreemptionSeconds is the time in seconds that the ClusterQueue
+  // is willing to wait for Workloads to finish before issuing preemptions,
+  // since the first time that the ClusterQueue has pending workloads that would
+  // fit under the min quota. This time is present in the timestamp for the
+  // PendingWorkloadsUnderMinQuota condition.
+  // If null, it means to wait indefinitely.
+  WaitBeforePreemptionSeconds *int64
+
+  // <UNRESOLVED>
+
+  // allowPreemptionByLowerPriority determines whether pending Workloads can
+  // preempt Workloads with higher priority in this ClusterQueue.
```

Comment:
```suggestion
  // preempt Workloads with lower priority in this ClusterQueue.
```

Reply:
I am missing something then, why would we want to allow lower priority workloads to preempt higher priority ones? or is this referring to higher priority workloads from other clusterQueues running on borrowed resources?

Reply:
Renaming to `AllowPreemptionByLowerPriorityWhenBorrowing` :)

Should we add this API? In a way, I'm thinking that high priority jobs should never run on borrowed capacity, which can be configured via quotas.

But an argument could be made that a high priority job probably should be able to use all the capacity available in a cluster.

Although, in both cases, it sounds like the default should be false and we can leave that as an implicit behavior for now.